packages/typegpu-react/src/use-mirrored-uniform.ts

```diff
@@ -0,0 +1,60 @@
+import * as d from 'typegpu/data';
+import { useRoot } from './root-context.tsx';
+import { useEffect, useMemo, useRef, useState } from 'react';
+import type { ValidateUniformSchema } from 'typegpu';
+
+interface MirroredValue<TSchema> {
+  schema: TSchema;
+  readonly $: d.InferGPU<TSchema>;
+}
+
+export function useMirroredUniform<
+  TSchema extends d.AnyWgslData,
+  TValue extends d.Infer<TSchema>,
+>(
+  schema: ValidateUniformSchema<TSchema>,
+  value: TValue,
+): MirroredValue<TSchema> {
+  const root = useRoot();
+  const [uniformBuffer, setUniformBuffer] = useState(() => {
+    return root.createUniform(schema, value);
+  });
+  const prevSchemaRef = useRef(schema);
+  const cleanupRef = useRef<ReturnType<typeof setTimeout> | null>(null);
+
+  useEffect(() => {
+    let currentBuffer = uniformBuffer;
+    if (!d.deepEqual(prevSchemaRef.current as d.AnyData, schema as d.AnyData)) {
+      currentBuffer.buffer.destroy();
+      currentBuffer = root.createUniform(schema, value);
+      setUniformBuffer(currentBuffer);
+      prevSchemaRef.current = schema;
+    }
+
+    currentBuffer.write(value);
+  }, [schema, value, root, uniformBuffer]);
+
+  useEffect(() => {
+    if (cleanupRef.current) {
+      clearTimeout(cleanupRef.current);
+    }
+
+    return () => {
+      cleanupRef.current = setTimeout(() => {
+        uniformBuffer.buffer.destroy();
+      }, 200);
+    };
+  }, [uniformBuffer]);
+
+  const mirroredValue = useMemo(
+    () => ({
+      schema,
+      get $() {
+        return uniformBuffer.$;
+      },
+    }),
+    [schema, uniformBuffer],
+  );
```

Comment:
This value is not stable (depends on a possible unstable `schema` value). I would suggest to keep a `currentSchemaRef` and update it during rendering only if the deepEquals no longer passes

```ts
const currentSchemaRef = useRef(schema);
if (!d.deepEqual(currentSchemaRef.current, schema)) {
  currentSchemaRef.current = schema;
}
```

And then use that value in the dependency array:

```ts
[currentSchemaRef.current, uniformBuffer]
```

You can add comments explaining why you're writing that kind of code, since this seems like betraying rules of React, but you gotta do what you gotta do 😅